### PR TITLE
Add support for .Net Core

### PR DESCRIPTION
**Pull request by PhenX**

### CODE REVIEW

**Comment by tikal** [comment]
@PhenX FYI, tests are failing

File: OpenXmlPowerTools.Tests/OpenXmlPowerTools.Tests.csproj [hunk 1]
```diff
@@ -1,28 +1,24 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>net452;net46</TargetFrameworks>
     <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
     <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
     <GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>
+    <TargetFrameworks>net461;netcoreapp2.0</TargetFrameworks>
```
Comment (twsouthwick) [inline OpenXmlPowerTools.Tests/OpenXmlPowerTools.Tests.csproj]: Why did you change the target frameworks? Let's keep net452 in the list as well - this allows testing the .net 4.5 build

File: OpenXmlPowerTools.Tests/OpenXmlPowerTools.Tests.csproj [hunk 2]
```diff
@@ -1,28 +1,24 @@
...
-    <PackageReference Include="xunit.runner.console" Version="2.3.1" />
+    <PackageReference Include="Microsoft.NET.Test.Sdk" Version="15.7.2" />
+    <PackageReference Include="xunit" Version="2.4.0" />
+    <PackageReference Include="xunit.runner.visualstudio" Version="2.4.0" />
+    <PackageReference Include="xunit.runner.console" Version="2.4.0" />
   </ItemGroup>
 
   <ItemGroup>
     <ProjectReference Include="..\OpenXmlPowerTools\OpenXmlPowerTools.csproj" />
   </ItemGroup>
```
Comment (twsouthwick) [inline OpenXmlPowerTools.Tests/OpenXmlPowerTools.Tests.csproj]: WindowsBase is needed for .NET 4.5. The SDK supports WindowsBase for <4.6, and >=4.6 uses System.IO.packaging

File: OpenXmlPowerTools/HtmlToWmlCssParser.cs [hunk 3]
```diff
@@ -1523,11 +1521,15 @@ public bool IsColor
                         return false;
                     }
 
-                    KnownColor kc;
-                    if (Enum.TryParse(m_val, true, out kc))
+                    try
                     {
+                        Color.FromName(m_val);
```
Comment (twsouthwick) [inline OpenXmlPowerTools/HtmlToWmlCssParser.cs]: Is there a way to do this without a try/catch?

File: OpenXmlPowerTools/OpenXmlPowerTools.csproj [hunk 4]
```diff
@@ -1,21 +1,10 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
   <PropertyGroup>
-    <TargetFrameworks>net45;net46</TargetFrameworks>
+    <TargetFrameworks>net461;netstandard2.0</TargetFrameworks>
```
Comment (twsouthwick) [inline OpenXmlPowerTools/OpenXmlPowerTools.csproj]: Please keep the same support and just add .NET Standard 2.0

File: OpenXmlPowerTools/PtOpenXmlUtil.cs [hunk 5]
```diff
@@ -678,13 +678,15 @@ public static int CalcWidthOfRunInTwips(XElement r)
             if (rPr == null)
                 throw new OpenXmlPowerToolsException("Internal Error, should have run properties");
             var languageType = (string)r.Attribute(PtOpenXml.LanguageType);
-            decimal? sz = null;
+            decimal? szn = null;
             if (languageType == "bidi")
-                sz = (decimal?)rPr.Elements(W.szCs).Attributes(W.val).FirstOrDefault();
+                szn = (decimal?)rPr.Elements(W.szCs).Attributes(W.val).FirstOrDefault();
```
Comment (twsouthwick) [inline OpenXmlPowerTools/PtOpenXmlUtil.cs]: I don't think this will work. W.val is a double, so default will end up being `0`.

File: OpenXmlPowerTools/PtOpenXmlUtil.cs [hunk 6]
```diff
@@ -743,66 +745,42 @@ public static int CalcWidthOfRunInTwips(XElement r)
...
-                            var twip = (int)(((sf.Width / dpi) * 1440m) / multiplier + tabLength * 1440m);
-                            return twip;
-                        }
+                        w = MetricsGetter.GetTextWidth(ff2, fs, sz, runText);
                     }
                 }
             }
+            catch (OverflowException)
+            {
+                return 0;
```
Comment (twsouthwick) [inline OpenXmlPowerTools/PtOpenXmlUtil.cs]: `MetricsGetter....` should handle all exceptions

File: OpenXmlPowerTools/PtOpenXmlUtil.cs (hunk 5 above)
Comment (PhenX) [reply OpenXmlPowerTools/PtOpenXmlUtil.cs]: This part of the code is the same as before, I only introduced a non nullable decimal after this.

File: OpenXmlPowerTools/PtOpenXmlUtil.cs (hunk 6 above)
Comment (PhenX) [reply OpenXmlPowerTools/PtOpenXmlUtil.cs]: I'll remove this catch because anyways GetTextWidth cacthes everything.

**Comment by PhenX** [comment]
I took your comments into account in the last commit, tell me if I forgot something :)

File: OpenXmlPowerTools/HtmlToWmlCssParser.cs (hunk 3 above)
Comment (PhenX) [reply OpenXmlPowerTools/HtmlToWmlCssParser.cs]: I didn't find one, I agree that that's far from ideal ... Maybe we could introduce a simple caching system with a dictionary ?

File: OpenXmlPowerTools/HtmlToWmlConverterCore.cs [hunk 7]
```diff
@@ -1232,55 +1231,28 @@ private static XElement FontMerge(XElement higherPriorityFont, XElement lowerPri
...
             {
-                using (Font f = new Font(ff, (float)sz / 2f, fs))
-                {
-                    const TextFormatFlags tff = TextFormatFlags.NoPadding;
-                    var proposedSize = new Size(int.MaxValue, int.MaxValue);
-                    var sf = TextRenderer.MeasureText(runText, f, proposedSize, tff);
-                    // sf returns size in pixels
-                    return sf.Width / multiplier;
-                }
+                return MetricsGetter.GetTextWidth(ff, fs, sz, runText) / multiplier;
```
Comment (twsouthwick) [inline OpenXmlPowerTools/HtmlToWmlConverterCore.cs]: Can you update this so that MetrixGetter.GetTextWidth always returns a value - any error checking is handled there?

File: OpenXmlPowerTools/HtmlToWmlCssParser.cs (hunk 3 above)
Comment (twsouthwick) [reply OpenXmlPowerTools/HtmlToWmlCssParser.cs]: Maybe something similar to `MetricsGetter` like `ColorParser` that uses the logic from betfore (ie `Enum.Parse` but instead does `Enum.TryParse<>`)

Comment (twsouthwick) [reply OpenXmlPowerTools/HtmlToWmlCssParser.cs]: And to handle a not named color, do a `bool ColorParser.TryGetColor(string value, out Color color)` kind of pattern

File: OpenXmlPowerTools/MetricsGetter.cs [hunk 8]
```diff
@@ -116,6 +117,26 @@ public static XElement GetDocxMetrics(WmlDocument wmlDoc, MetricsGetterSettings
...
+            try
+            {
+                using (var f = new Font(ff, (float)sz / 2f, fs))
+                {
+                    var proposedSize = new Size(int.MaxValue, int.MaxValue);
+                    var sf = graphics.MeasureString(text, f, proposedSize);
+                    return (int) sf.Width;
+                }
+            }
+            catch
```
Comment (twsouthwick) [inline OpenXmlPowerTools/MetricsGetter.cs]: There's a fall back that the code originally had to use Times new Roman, right?

File: OpenXmlPowerTools/OpenXmlPowerTools.csproj [hunk 9]
```diff
@@ -1,21 +1,24 @@
...
   <PropertyGroup>
-    <TargetFrameworks>net45;net46</TargetFrameworks>
+    <TargetFrameworks>net45;net46;netstandard2.0</TargetFrameworks>
   </PropertyGroup>
 
   <ItemGroup>
     <PackageReference Include="DocumentFormat.OpenXml" Version="2.8.1" />
   </ItemGroup>
-
+
```
Comment (twsouthwick) [inline OpenXmlPowerTools/OpenXmlPowerTools.csproj]: can you remove the extra whitespace?

File: OpenXmlPowerTools/PtOpenXmlUtil.cs (hunk 5 above)
Comment (twsouthwick) [reply OpenXmlPowerTools/PtOpenXmlUtil.cs]: ah, you're right. It's a bug, but not a regression :)

Comment (twsouthwick) [reply OpenXmlPowerTools/PtOpenXmlUtil.cs]: Can you create an issue to track this?

File: OpenXmlPowerTools/PtOpenXmlUtil.cs [hunk 10]
```diff
@@ -743,66 +745,38 @@ public static int CalcWidthOfRunInTwips(XElement r)
                 runText = sb.ToString();
             }
 
+            int w;
```
Comment (twsouthwick) [inline OpenXmlPowerTools/PtOpenXmlUtil.cs]: The next block seems like it could be moved into the `MetricGetter.GetTextWidth` with an appropriate default return value (ie, as mentioned above, handle all possible errors there)

File: OpenXmlPowerTools/WmlToHtmlConverter.cs [hunk 11]
```diff
@@ -2370,66 +2369,33 @@ private static int CalcWidthOfRunInTwips(XElement r)
...
-                {
-                    const TextFormatFlags tff = TextFormatFlags.NoPadding;
-                    var proposedSize = new Size(int.MaxValue, int.MaxValue);
-                    var sf = TextRenderer.MeasureText(runText, f, proposedSize, tff);
-                        // sf returns size in pixels
-                    const decimal dpi = 96m;
-                    var twip = (int) (((sf.Width/dpi)*1440m)/multiplier + tabLength*1440m);
-                    return twip;
-                }
+                w = MetricsGetter.GetTextWidth(ff, fs, sz, runText);
```
Comment (twsouthwick) [inline OpenXmlPowerTools/WmlToHtmlConverter.cs]: As above, all errors (and fallbacks), should be handled in the `GetTextWidth` method

**Comment by PhenX** [comment]
Is it ok now ?

**Comment by PhenX** [comment]
Also, do you know when the Nuget package is able to be updated with these changes ?

**Comment by vladimir-shmidt** [comment]
Hi guys, any news about this? appveyor says that one test have been failed: 
`[xUnit.net 00:00:04.23]     OxPt.PbTests.PB006_VideoFormats [FAIL]`
Error Message:
 System.IO.IOException : Entries cannot be opened multiple times in Update mode.
This error looks very the same as https://github.com/dotnet/corefx/issues/26952 and was fixed and merged in https://github.com/dotnet/corefx/pull/28478/commits/8797ecbffd8f050cb6cadff8f790fac6faf83bed

**Comment by vladimir-shmidt** [comment]
Hi guys, any news about this? appveyor says that one of the test have been failed: 
`[xUnit.net 00:00:04.23]     OxPt.PbTests.PB006_VideoFormats [FAIL]`
Error Message:
 System.IO.IOException : Entries cannot be opened multiple times in Update mode.
This error looks very the same as https://github.com/dotnet/corefx/issues/26952 and was fixed and merged in https://github.com/dotnet/corefx/pull/28478/commits/8797ecbffd8f050cb6cadff8f790fac6faf83bed

**Comment by twsouthwick** [comment]
@vladimir-shmidt Sorry for the delay - been swamped with stuff. Thanks for the fixes - I'll go through it again.

As far as the error, go ahead and change the tests to target .NET Core 2.1 and see if that fixes it. It's not ideal, as we should fix the tests to work correctly. I saw this often on the OpenXml SDK where streams were not being disposed of correctly. So, those are two options to fix it and I'll leave it up to you - I prefer fixing disposing if that's the issue, but if it's truly a .NET bug, then we won't worry about it.

File: OpenXmlPowerTools.Tests/OpenXmlPowerTools.Tests.csproj [hunk 12]
```diff
@@ -1,24 +1,24 @@
 ﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
-    <TargetFrameworks>net452;net46</TargetFrameworks>
     <AutoGenerateBindingRedirects>true</AutoGenerateBindingRedirects>
     <GenerateBindingRedirectsOutputType>true</GenerateBindingRedirectsOutputType>
     <GenerateRuntimeConfigurationFiles>true</GenerateRuntimeConfigurationFiles>
+    <TargetFrameworks>net452;net461;netcoreapp2.0</TargetFrameworks>
```
Comment (twsouthwick) [inline OpenXmlPowerTools.Tests/OpenXmlPowerTools.Tests.csproj]: nit: can you move this bag to the top - makes it easier to identify.

File: OpenXmlPowerTools.Tests/WmlComparerTests.cs [hunk 13]
```diff
@@ -171,7 +171,7 @@ public void WC001_Consolidate(string testId, string originalName, string revised
                     return new WmlRevisedDocumentInfo()
                     {
                         RevisedDocument = new WmlDocument(revisedCopiedToDestDocx.FullName),
-                        Color = Color.FromName(z.Element("Color").Value),
+                        Color = ColorParser.FromName(z.Element("Color").Value),
```
Comment (twsouthwick) [inline OpenXmlPowerTools.Tests/WmlComparerTests.cs]: What about if the element doesn't exist? Is this verified before? Seems like some null checking is needed. It would be easiest to do `z.Element("Color")?.Value` and then do the null check in the parser.

This applies to all the similar changes below.

File: OpenXmlPowerTools/ColorParser.cs [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+Twitter: @EricWhiteDev
+Email: eric@ericwhite.com
+
+***************************************************************************/
+
+using System.Drawing;
+
+namespace OpenXmlPowerTools
+{
+    public static class ColorParser
```
Comment (twsouthwick) [inline OpenXmlPowerTools/ColorParser.cs]: What about error checking in here for exceptions?

File: OpenXmlPowerTools/HtmlToWmlCssParser.cs [hunk 15]
```diff
@@ -835,13 +833,11 @@ public Color ToColor()
             }
             else
             {
-                try
+                Color c;
+                if (ColorParser.TryFromName(m_value, out c))
```
Comment (twsouthwick) [inline OpenXmlPowerTools/HtmlToWmlCssParser.cs]: `out var c`

File: OpenXmlPowerTools/HtmlToWmlCssParser.cs [hunk 16]
```diff
@@ -1649,13 +1644,11 @@ public Color ToColor()
             }
             else
             {
-                try
+                Color c;
+                if (ColorParser.TryFromName(m_val, out c))
```
Comment (twsouthwick) [inline OpenXmlPowerTools/HtmlToWmlCssParser.cs]: `out var c`

File: OpenXmlPowerTools/MetricsGetter.cs [hunk 17]
```diff
@@ -39,6 +40,21 @@ public class MetricsGetterSettings
 
     public class MetricsGetter
     {
+        private static Graphics _graphics;
+        protected static Graphics Graphics
```
Comment (twsouthwick) [inline OpenXmlPowerTools/MetricsGetter.cs]: Should this be private? What about using `Lazy<Graphics>` for this?

**Comment by twsouthwick** [comment]
Ah - I misunderstood the issue you linked to. Looks like this is fixed for all versions, just in a later `System.IO.Packaging`. v2.9.0 of the SDK brings in the latest changes and should have it fixed. However, we've run into some issues putting it on the official nuget stream - it's only on the CI. You can update to v2.9.0 of the SDK and it'll be fixed, just need to add an entry in the NuGet.config for `https://dotnet.myget.org/F/open-xml-sdk/api/v3/index.json`

**Comment by vladimir-shmidt** [comment]
@twsouthwick would it better to wait for @PhenX to update this PR or start new session with link to this?

**Comment by PhenX** [comment]
Hello guys, I'm on vacation this week, I'll apply the changes next week :)

**Comment by twsouthwick** [comment]
Sounds like @PhenX is on this - sorry for the slow responses. I meant to tell @PhenX thanks for the fixes - sorry for the confusion. Have a good vacation and I'll be watching for any updates here so we can get it merged in.

File: OpenXmlPowerTools.Tests/WmlComparerTests.cs (hunk 13 above)
Comment (PhenX) [reply OpenXmlPowerTools.Tests/WmlComparerTests.cs]: @twsouthwick Do you want me to change code that was like this before ? I think it's better to keep the code external to my PR as it is and handle it in another PR, what do you think?

File: OpenXmlPowerTools/HtmlToWmlCssParser.cs (hunk 15 above)
Comment (PhenX) [reply OpenXmlPowerTools/HtmlToWmlCssParser.cs]: It's a C# 7.0 feature, the project is currently on C# 4.0 (https://github.com/OfficeDev/Open-Xml-PowerTools/blob/vNext/OpenXmlPowerTools/OpenXmlPowerTools.csproj.DotSettings) Do you confirm I should change this ? I'm totally ok with that but wanted to be sure :)

File: OpenXmlPowerTools/ColorParser.cs [hunk 18]
```diff
@@ -29,14 +29,30 @@ public static Color FromName(string name)
...
-            return color.IsNamedColor;
+            try
+            {
+                color = Color.FromName(name);
+
+                return color.IsNamedColor;
+            }
+            catch
+            {
+                color = new Color();
```
Comment (twsouthwick) [inline OpenXmlPowerTools/ColorParser.cs]: Does this need to be set to an instance? Why not set it to null? If it's a struct, use `default(Color)`

File: OpenXmlPowerTools/ColorParser.cs [hunk 19]
```diff
@@ -29,14 +29,30 @@ public static Color FromName(string name)
...
+                return false;
+            }
         }
 
         public static bool IsValidName(string name)
         {
-            return Color.FromName(name).IsNamedColor;
+            try
+            {
+                return Color.FromName(name).IsNamedColor;
```
Comment (twsouthwick) [inline OpenXmlPowerTools/ColorParser.cs]: Use `TryFromName` here

File: OpenXmlPowerTools/MetricsGetter.cs [hunk 20]
```diff
@@ -40,18 +40,14 @@ public class MetricsGetterSettings
 
     public class MetricsGetter
     {
-        private static Graphics _graphics;
-        protected static Graphics Graphics
-        {
+        protected static Lazy<Graphics> Graphics {
```
Comment (twsouthwick) [inline OpenXmlPowerTools/MetricsGetter.cs]: does this need to be `protected`? Also, make it `readonly`

File: OpenXmlPowerTools.Tests/WmlComparerTests.cs (hunk 13 above)
Comment (twsouthwick) [reply OpenXmlPowerTools.Tests/WmlComparerTests.cs]: Can you at least add appropriate null checking where you're touching code? I'm fine doing it in another PR, though, if you feel it will be a lot of extra noise. If you go that route, open a new issue and link it here.

File: OpenXmlPowerTools/HtmlToWmlCssParser.cs (hunk 15 above)
Comment (twsouthwick) [reply OpenXmlPowerTools/HtmlToWmlCssParser.cs]: Yup. I didn't realize it was still on 4.0! Go ahead and update it.

File: OpenXmlPowerTools/MetricsGetter.cs [hunk 21]
```diff
@@ -39,6 +40,17 @@ public class MetricsGetterSettings
 
     public class MetricsGetter
     {
+        protected static Lazy<Graphics> Graphics {
```
Comment (twsouthwick) [inline OpenXmlPowerTools/MetricsGetter.cs]: Does this need to be `protected` instead of `private`? Also, make it `readonly`

**Comment by twsouthwick** [comment]
Looks like the test is still failing. Are you going to update the `System.IO.Packaging` required?

File: OpenXmlPowerTools/MetricsGetter.cs (hunk 21 above)
Comment (PhenX) [reply OpenXmlPowerTools/MetricsGetter.cs]: I made it private, but no need to make it readonly as it only has a getter.

**Comment by PhenX** [comment]
@twsouthwick I could not find a way to update `System.IO.Packaging` by including a NuGet package. It doesn't work better.
I think we would need to upgrade to netcoreapp2.1 (as you said in your first answser, it's a bug in NetCore 2.0), do you have any other suggestion ?

File: OpenXmlPowerTools/MetricsGetter.cs (hunk 21 above)
Comment (twsouthwick) [reply OpenXmlPowerTools/MetricsGetter.cs]: Oh I didn't see that. You're actually creating a new lazy each time :) Instead, do this:

```csharp
private static Lazy<Graphics> Graphics { get; } = new Lazy<Graphics>(()=>...)
```

Comment (PhenX) [reply OpenXmlPowerTools/MetricsGetter.cs]: Haha ! Good catch

**Comment by twsouthwick** [comment]
Go ahead and target 2.1 for the tests - keep the library as .NET STandard 2.0, though

File: OpenXmlPowerTools/MetricsGetter.cs [hunk 22]
```diff
@@ -40,16 +40,11 @@ public class MetricsGetterSettings
...
-                return new Lazy<Graphics>(() =>
-                {
-                    Image image = new Bitmap(1, 1);
-                    return System.Drawing.Graphics.FromImage(image);
-                });
-            }
-        }
+        private static Lazy<Graphics> Graphics { get; } = new Lazy<Graphics>(() =>
+        {
+            Image image = new Bitmap(1, 1);
```
Comment (twsouthwick) [inline OpenXmlPowerTools/MetricsGetter.cs]: Just wondering now, how expensive is the graphics invocation? Does it need to be lazy? I'm fine with how it is, but just wondered if any perf testing was done to see if this was helpful?

Comment (PhenX) [reply OpenXmlPowerTools/MetricsGetter.cs]: No I didn't check, but instinctively this kind of trick takes a lot of time and memory if repeated a lot.

**Comment by PhenX** [comment]
Well, it still doesn't work, but it looks like @EricWhiteDev has had this problem once http://www.ericwhite.com/blog/forums/topic/system-io-ioexception-after-upgrade-to-openxml-sdk-2-6/

**Comment by twsouthwick** [comment]
Hmmm just so we can get this merged in for now, can you add a skip attribute to that test and open an issue?

**Comment by PhenX** [comment]
Just added the skip attribute, I'll open an issue once merged :)

**Comment by twsouthwick** [comment]
Cool. Thanks for getting this support! I'll merge this in, and we can deal with the outstanding issues in here later

**Comment by PhenX** [comment]
Great! I'll open the issue quickly. When do you think a new version of the nuget package can be released?

**Comment by PhenX** [comment]
@twsouthwick sorry to insist :blush:, when do you think a new version of the nuget package can be released?

**Comment by twsouthwick** [comment]
You can currently access the NuGet via the AppVeyor feed. We're working through some policy changes due to NuGet package signing for the Open XML SDK and will try to include this library in that so we can publish it to NuGet.

**Comment by augb** [comment]
> You can currently access the NuGet via the AppVeyor feed. ...

Is this publicly available? If so, what is the feed URL?

**Comment by twsouthwick** [comment]
@augb It is public, but apparently wasn't documented. I've added it to the README

**Comment by augb** [comment]
Thank you @twsouthwick. There don't appear to be any packages: https://ci.appveyor.com/nuget/open-xml-powertools. When I use the `dotnet` cli command, I get a `500 - Internal Error`:
```
dotnet add package OpenXmlPowerTools --source https://ci.appveyor.com/nuget/open-xml-powertools/api/v2/package
... elided ...
info :   InternalServerError https://ci.appveyor.com/nuget/open-xml-powertools/api/v2/package/FindPackagesById()?id='OpenXmlPowerTools'&semVerLevel=2.0.0 87ms
error: Failed to retrieve information about 'OpenXmlPowerTools' from remote source 'https://ci.appveyor.com/nuget/open-xml-powertools/api/v2/package/FindPackagesById()?id='OpenXmlPowerTools'&semVerLevel=2.0.0'.
error:   Response status code does not indicate success: 500 (Invalid NuGet feed request.).
```

**Comment by twsouthwick** [comment]
Hmmm that does appear to be the case. I'll look into that.

**Comment by rmourato** [comment]
Looking forward to an official Nuget package with these changes in!

**Comment by PhenX** [comment]
@EricWhiteDev @twsouthwick Sorry to bother you again, but I think the compatibility with NetStandard is an important point for the community, and having a NuGet package with the latest changes (including NetStandard support) is something a lot of people, including me, are waiting for.
Thank you very much for your time taking this into account.
